doc/design/block.md

```diff
@@ -0,0 +1,128 @@
+# Block design
+
+In computer programming, a block is a lexical structure of source code which is grouped together.
+In most programming languages, block is useful when define a function or some conditional statements such as `if` and `while`.
+
+In Paddle, we need a similar concept to support following scenes:
+
+- the basic unit of execution, just like the original `NetOp`, block can group multiple operators and make them behave just like a single operator. Unlike `NetOp`, a network should not be an operator, but a block can be treated as a operator.
+  - Block should have a `Exec` method, which will execute the operators grouped in the block.
+- store the context for operators, including the information of variables they will operator on. 
+- Support the definition and execution of special operators such as `RNNOp`, `switch_op`, `while_op`, `if_else_op`, which needs one or more code-blocks.
+  - Block will help to define and execute thouse operators.
+  
+## when need a block
+Blocks are needed in following code snippets
+```c++
+if(cond) { block_true } else { block_false }
+```
+
+```c++
+while(cond){ block }
+```
+
+```c++
+int func(inputs) { block with outputs }
+```
+
+```c++
+namespace xxxx {
+  block
+}
+```
+
+Paddle will have `if_else_op`, `switch_op`, `while_op` and some other operators that need a function block such as RNNOp, and enable definition and execution of a block is vital to those operators.
+
+What's more, Paddle should has a `pd.namespace`, which will help to preventing name conflicts of variables in large model.
```

Comment:
We'd also like to introduce a namespace API for PaddlePaddle, which likes the C++ namespace, names a Block and prefixes the names of variables defined in it.